pandas/_libs/tslibs/timedeltas.pyx

```diff
@@ -506,6 +507,52 @@ def _binary_op_method_timedeltalike(op, name):
 # ----------------------------------------------------------------------
 # Timedelta Construction
 
+iso_pater = re.compile(r"""P
+                        (?P<days>-?[0-9]*)DT
+                        (?P<hours>[0-9]{1,2})H
+                        (?P<minutes>[0-9]{1,2})M
+                        (?P<seconds>[0-9]{0,2})
+                        (\.
+                        (?P<milliseconds>[0-9]{1,3})
+                        (?P<microseconds>[0-9]{0,3})
+                        (?P<nanoseconds>[0-9]{0,3})
+                        )?S""", re.VERBOSE)
+
+
+cdef int64_t parse_iso_format_string(object iso_fmt):
+    """
+    Extracts and cleanses the appropriate values from a match object with
+    groups for each component of an ISO 8601 duration
+
+    Parameters
+    ----------
+    iso_fmt:
+        ISO 8601 Duration formatted string
+
+    Returns
+    -------
+    ns: int64_t
+        Precision in nanoseconds of matched ISO 8601 duration, or -1 if
+        `iso_fmt` cannot be parsed
+    """
+
+    cdef int64_t ns = 0
+
+    match = re.match(iso_pater, iso_fmt)
+    if match:
+        match_dict = match.groupdict(default='0')
+        for comp in ['milliseconds', 'microseconds', 'nanoseconds']:
+            match_dict[comp] = '{:0<3}'.format(match_dict[comp])
+
+        for k, v in match_dict.items():
+            ns += timedelta_from_spec(v, '0', k)
+
+    else:
```

Comment:
you should just raise here

Reply:
Not an expert on Cython but pytest was failing when raising the exception inside of this function. I believe it is attributable to the int64_t function declaration. I could remove that type and raising the exception directly inside the function would work. Otherwise, I was getting the below output during testing.

```
-------------------------------------------------------------- Captured stderr call --------------------------------------------------------------
AttributeError: 'NoneType' object has no attribute 'groupdict'
Exception ignored in: 'pandas._libs.tslibs.timedeltas.parse_iso_format_string'
AttributeError: 'NoneType' object has no attribute 'groupdict'
```

Reply:
oh, you just need to add:
 ``except? -1:`` to the declaration; this tells cython that you may raise in a ``cdef`` function so it should check. you don't need to explicity return -1 though.

Reply:
Ah OK thanks. Will make changes and re-push